modules/programs/joplin-desktop.nix

```diff
@@ -0,0 +1,129 @@
+{ config, lib, pkgs, ... }:
+let
+  cfg = config.programs.joplin-desktop;
+  # config path is the same for linux and mac
+  configPath = (config.home.homeDirectory + "/.config/joplin-desktop");
+in {
+  options.programs.joplin-desktop = {
+    enable = lib.mkEnableOption "joplin-desktop";
+    package = lib.mkOption {
+      type = lib.types.package;
+      default = pkgs.joplin-desktop;
+      defaultText = lib.literalExpression "pkgs.joplin-desktop";
+      description = "Package containing the joplin-desktop program.";
+    };
+    extraConfig = lib.mkOption {
+      type = lib.types.attrs;
+      default = { };
+      description = ''
+        Use this to add options to the Joplin config file. Settings are written in JSON, so "sync.interval":600 is written as "sync.interval" = 600'';
+      example = {
+        "newNoteFocus" = "title";
+        "markdown.plugin.mark" = true;
+      };
+    };
+    ### General
+    general = {
+      editor = lib.mkOption {
+        type = lib.types.nullOr lib.types.str;
+        default = null;
+        example = "kate";
+        description =
+          "The editor command (may include arguments) that will be used to open a note. If none is provided Joplin will try to auto-detect the default editor.";
+      };
+    };
+    ### Sync
+    sync = {
+      target = lib.mkOption {
+        type = lib.types.enum [
+          null
+          "none"
+          "file-system"
+          "onedrive"
+          "nextcloud"
+          "webdav"
+          "dropbox"
+          "s3"
+          "joplin-server"
+          "joplin-cloud"
+        ];
+        default = null;
+        description = "What is the type of sync target.";
+        example = "dropbox";
+      };
+      interval = lib.mkOption {
+        type = lib.types.nullOr lib.types.int;
+        default = null;
+        description = ''
+          Sync Interval in seconds. Only values shown as options in the settings are valid.
+          Disabled: 0
+          5 Min: 300
+          10 Min: 600
+          30 Min: 1800
+          1 hour: 3600
+          12 Hours: 43200
+          24 Hours: 86400
```

Comment:
Not sure I understand the description, only 300, 600, etc. are valid values for the interval? If so, then I would suggest using an enum:

``` nix
type = types.enum [ 300 600 1800 3600 43200 86400 ];
```

or even

``` nix
type = types.enum [ "5m" "10m" "30m" "1h" "12h" "1d" ];
```

Reply:
These are the values that are available via the GUI, from my testing it does not seem like other values can be used. When using `type = types.enum [ "5m" "10m" "30m" "1h" "12h" "1d" ]`, we need some way to convert that into a number of seconds to write to the config file and I didn't find an elegant way to do so.

Reply:
I changes it from 
`type = lib.types.nullOr lib.types.int;`
to 
`type = lib.types.nullOr type = types.enum [ 300 600 1800 3600 43200 86400 ];`

Reply:
The typical way to do such a mapping would be something like

``` nix
intervalToSeconds = i:
  {
    "5m"  =   300;
    "10m" =   600;
    "30m" =  1800;
    "1h"  =  3600;
    "12h" = 43200;
    "1d"  = 86400;
  }.${i} or null;
```

Example:

```
nix-repl> intervalToSeconds 4
error:
       … while evaluating an attribute name

         at «none»:0: (source not available)

       error: value is an integer while a string was expected

nix-repl> intervalToSeconds "sdf"
null

nix-repl> intervalToSeconds "5m"  
300

nix-repl> intervalToSeconds "1d" 
86400
```

Reply:
Thanks a lot for that, I've used that now for the sync target as well.